bioconda_utils/update_pinnings.py

```diff
@@ -1,88 +1,125 @@
+
 buildNumberRegex = re.compile("^( +number: +)(\d+)(.*?)$")
 buildNumberJinja = re.compile("^({% +set build +=.+?)(\d+)(.*?%})")
 
-def already_bumped(r, name, version, buildNumber):
-    rsub = r.loc[(r.name == name) &
-             (r.platform != "osx") &
-             (r.version == version)]
-    if rsub.shape[0] == 0:
-        # The version was never built, it doesn't explicitly require bumping
-        return True
-    if buildNumber > max(rsub.build_number):
-        # Already bumped
-        return True
-    return False
 
+def will_build_variant(meta):
+    """Check if the recipe variant will be built as currently rendered"""
+    build_numbers = RepoData().get_package_data(
+        'build_number',
+        name=meta.name(), version=meta.version(),
+        platform=['linux', 'noarch'],
+    )
+    res = all(num < meta.build_number() for num in build_numbers)
+    if res:
+        logger.debug("Package %s=%s will be built already because %s < %s)",
+                     meta.name(), meta.version(),
+                     max(build_numbers) if build_numbers else "N/A",
+                     meta.build_number())
+    return res
 
-def already_built(r, name, version, buildstring):
-    if r.loc[(r.name == name) &
-             (r.platform != "osx") &
-             (r.version == version) &
-             (r.build == buildstring)].shape[0] > 0:
-        return True
-    return False
 
+def have_variant(meta):
+    """Checks if we have an exact match to name/version/buildstring
+    """
+    res = RepoData().get_package_data(
+        name=meta.name(), version=meta.version(), build=meta.build_id(),
+        platform=['linux', 'noarch']
+    )
+    if res:
+        logger.debug("Package %s=%s=%s exists",
+                     meta.name(), meta.version(), meta.build_id())
+    return res
 
-def only_python(r, name, version, buildstring):
-    s = r.loc[(r.name == name) &
-              (r.platform != "osx") &
-              (r.version == version)]
-    if s.shape[0] == 0:
-        # Version doesn't exist in linux/noarch
-        return False
-    build_set = set()
-    for b in s.build:
-        bs = b.split("_")
-        if bs[0].startswith("py"):
-            build_set.add(bs[0][4:])
-        else:
-            build_set.add(bs[0])
-    if buildstring in build_set:
-        return True
-    # Version exists, but build doesn't
-    return False
 
+def have_variant_but_for_python(meta):
+    """Checks if we have an exact or `py[23]_` prefixed match to
+    name/version/buildstring
 
-def should_be_bumped(recipe, build_config, repodata):
+    Ignores osx.
     """
-    Determine if a given recipe should have its build number increments
-    (bumped) due to a recent change in pinnings.
+    def strip_py(build):
+        if build.startswith("py"):
+            return build[4:]
+        return build
+
+    builds = RepoData().get_package_data(
+        'build',
+        name=meta.name(), version=meta.version(),
+        platform=['linux', 'noarch']
```

Comment:
@dpryan79 - Why did you exclude OSX from the consideration?

Reply:
In practice everything in OSX is in Linux (with few if any exceptions) and the pinnings are almost entirely identical, so I figured I'd save the time. We can add OSX in, it'll just increase the run time slightly. I guess if someone were to run this on OSX it would currently incorrectly think that almost everything should be updated, so I suppose that's a compelling argument for adding `osx` to that list of platforms.